src/pulsar_client_manager.erl

```diff
@@ -0,0 +1,381 @@
+                    %% Replying error to avoid loop
+                    {{error, pulsar_unstable}, State0};
+                {redirect, OtherServiceURL, OtherOpts} ->
+                    handle_redirect_lookup(State0, OtherServiceURL, OtherOpts, PartitionTopic);
+                Res ->
+                    {Res, State0}
+            catch
+                exit:{timeout, _} ->
+                    {{error, timeout}, State0};
+                exit:{noproc, _} ->
+                    %% race; retry
+                    timer:sleep(10),
+                    handle_redirect_lookup(State0, ServiceURL, Opts, PartitionTopic)
+            end;
+        error ->
+            maybe
+                {{ok, _Pid}, State} ?= spawn_any_and_wait_connected(State0, [ServiceURL]),
+                handle_redirect_lookup(State, ServiceURL, Opts, PartitionTopic)
+            end
+    end.
+
+handle_worker_down(State0, Pid, Reason) ->
+    log_info("worker ~0p down: ~0p", [Pid, Reason]),
+    #{?workers := Workers0} = State0,
+    {URL, Workers1} = maps:take(Pid, Workers0),
+    Workers = maps:remove(URL, Workers1),
+    State0#{?workers := Workers}.
+
+find_alive_worker(State, URL) ->
+    #{?workers := Workers} = State,
+    URLKey = bin(URL),
+    maybe
+        #{URLKey := Pid} ?= Workers,
+        true ?= is_process_alive(Pid),
+        {ok, Pid}
+    else
+        _ -> error
+    end.
+
+alive_workers(State) ->
+    #{?workers := Workers} = State,
+    Keys = maps:keys(Workers),
+    lists:filter(fun(Key) -> is_pid(Key) andalso is_process_alive(Key) end, Keys).
+
+get_first_successful_call(WorkerPids, Fun, Args) ->
+    pulsar_utils:foldl_while(
+      fun(WorkerPid, Acc) ->
+        try
+            {halt, apply(Fun, [WorkerPid | Args])}
+        catch
+            _:_ ->
+                {cont, Acc}
+        end
+      end,
+      {error, ?no_servers_available},
+      WorkerPids).
+
+%% Takes the list of seed URLs and attempts to have at most one worker alive and connected
+%% to it.
+-spec spawn_any_and_wait_connected(state()) -> {{ok, pid()} | {error, term()}, state()}.
+spawn_any_and_wait_connected(State0) ->
+    #{?seed_urls := SeedURLs} = State0,
+    spawn_any_and_wait_connected(State0, SeedURLs).
+
+spawn_any_and_wait_connected(State0, SeedURLs) ->
+    #{ ?client_id := ClientId
+     , ?initial_opts := Opts
+     , ?workers := Workers0
+     } = State0,
+    Res =
+        pulsar_utils:foldl_while(
+          fun(URL, _Acc) ->
+            case pulsar_client:start_link(ClientId, URL, Opts) of
+                {error, _} = Error ->
+                    {cont, Error};
+                {ok, Pid} ->
+                    {halt, {ok, {URL, Pid}}}
+            end
+          end,
+          {error, ?no_servers_available},
+          SeedURLs),
+    %% We currently don't use `ignore'.
+    case Res of
+        {ok, {URL, Pid}} ->
+            Workers = Workers0#{bin(URL) => Pid, Pid => true},
```

Comment:
The map is not very big, just find it in the map when EXIT is received.
```suggestion
            Workers = Workers0#{bin(URL) => Pid},
```

Reply:
:checkered_flag: